components/cam/src/physics/cam/shoc.F90

```diff
@@ -924,6 +893,130 @@ subroutine update_prognostics_implicit( &
+  do klev = 2, nlevi
+    do icol = 1, shcol
+       tmpi(icol,klev) = dtime * (ggr*rho_zi(icol,klev)) / dz_zi(icol,klev)
+    enddo
+  enddo
+
+end subroutine compute_tmpi
+
+subroutine dp_inverse(nlev, nlevi, shcol, rho_zt, dz_zt, rdp_zt)
+
+  !intent-ins
+  integer,     intent(in) :: nlev, nlevi, shcol
+  ! Air density on thermo grid [kg/m3]
+  real(rtype), intent(in) :: rho_zt(shcol,nlev)
+  ! height thickness centered on thermo grid [m]
+  real(rtype), intent(in) :: dz_zt(shcol,nlev)
+
+  !intent-out
+  real(rtype), intent(out) :: rdp_zt(shcol,nlev)
+
+  !local vars
+  integer :: icol, klev
+
+  do klev = 1, nlev
+    do icol = 1, shcol
+
+      rdp_zt(icol,klev) = 1._rtype/(ggr*rho_zt(icol,klev)*dz_zt(icol,klev))
+
+    enddo
+  enddo
+
+end subroutine dp_inverse
+
+subroutine compute_sfc_terms(nlev, nlevi, shcol, dtime, rho_zi, uw_sfc, vw_sfc, u_wind, &
+       v_wind, rdp_zt, wthl_sfc, wqw_sfc, thetal, qw, tke, ksrf)
+
+  !intent-ins
+  integer,     intent(in) :: nlev, nlevi, shcol
+
+  !time step [s]
+  real(rtype), intent(in) :: dtime
+  !air density at interfaces [kg/m3]
+  real(rtype), intent(in) :: rho_zi(shcol,nlevi)
+  !vertical zonal momentum flux at surface [m3/s3]
+  real(rtype), intent(in) :: uw_sfc(shcol)
+  !vertical meridional momentum flux at surface [m3/s3]
+  real(rtype), intent(in) :: vw_sfc(shcol)
+  !zonal wind [m/s]
+  real(rtype), intent(in) :: u_wind(shcol,nlev)
+  !meridional wind [m/s]
+  real(rtype), intent(in) :: v_wind(shcol,nlev)
+  !inverse of dp
+  real(rtype), intent(in) :: rdp_zt(shcol,nlev)
+  !vertical heat flux at surface [K m/s]
+  real(rtype), intent(in) :: wthl_sfc(shcol)
+  !vertical moisture flux at surface [kg/kg m/s]
+  real(rtype), intent(in) :: wqw_sfc(shcol)
+
+  !intent-inouts
+  ! liquid water potential temperature [K]
+  real(rtype), intent(inout) :: thetal(shcol,nlev)
+  ! total water mixing ratio [kg/kg]
+  real(rtype), intent(inout) :: qw(shcol,nlev)
+  ! turbulent kinetic energy [m2/s2]
+  real(rtype), intent(inout) :: tke(shcol,nlev)
+
+  !intent-out
+  real(rtype), intent(out) :: ksrf(shcol)
+
+  !local vars
+  integer :: icol
+
+  real(rtype) :: taux, tauy !stresses (N/m2)
+  real(rtype) :: ws         !wind speed (m/s)
+  real(rtype) :: rho, tau, ustar, fac, rdp, uw, vw, wtke_flux
+
+  real(rtype), parameter :: wsmin   = 1._rtype    ! Minimum wind speed for ksrfturb computation [ m/s ]
+  real(rtype), parameter :: ksrfmin = 1.e-4_rtype ! Minimum surface drag coefficient  [ kg/s/m^2 ]
+
+  do icol=1,shcol
+     rho        = rho_zi(icol,nlevi)
+     uw         = uw_sfc(icol)
+     vw         = vw_sfc(icol)
+
+     taux       = rho*uw ! stress in N/m2
+     tauy       = rho*vw ! stress in N/m2
+     ! compute the wind speed
+     ws         = max(sqrt(u_wind(icol,nlev)**2._rtype + v_wind(icol,nlev)**2._rtype), wsmin)
+     tau        = sqrt( taux**2._rtype + tauy**2._rtype )
+     ksrf(icol) = max(tau/ws, ksrfmin)
```

Comment:
@bogensch : I am working on further refactoring this code. It seems like `ksrf` variable is computed here and then again supplied to `vd_shoc_decomp` as an intent(out). It is never used again. Let me know if I am not interpreting it correctly.
If you think it is not used and we should remove it, it will also help in removing few more computations. Let me know and I will do what you suggest in my ongoing refactor.

Reply:
I'm not sure I follow.  On my end it appears that ksrf is identified as an intent(in) (variable "flux") in vd_shoc_decomp, which is needed for the surface fluxes for momentum.  It looks like the way you have it in your current implementation appears correct to me.

Reply:
Thanks for looking at the code. You are right, it is intent-in, I misread the code. I am almost done with your suggested refactor and will push my changes soon.